packages/apollo-gateway/src/__tests__/executeQueryPlan.test.ts

```diff
@@ -672,4 +672,50 @@ describe('executeQueryPlan', () => {
       }
     `);
   });
+
+  it('can execute queries with selections on unresolved @requires fields', async () => {
+    // query a book not known to the book service
+    const query = gql`
+      query {
+        products(upcs: ["9999999999", "0987654321"]) {
+          upc
+          name
+          price
+        }
+      }
+    `;
+
+    const operationContext = buildOperationContext(schema, query);
+    const queryPlan = buildQueryPlan(operationContext);
+
+    const response = await executeQueryPlan(
+      queryPlan,
+      serviceMap,
+      buildRequestContext(),
+      operationContext,
+    );
+
+    expect(response.errors?.map(e => e.message)).toEqual(expect.arrayContaining([
+      "Cannot return null for non-nullable field Book.isbn.",
+      "Field \"title\" was not found in response.",
+      "Field \"year\" was not found in response.",
+    ]));
+
+    expect(response.data).toMatchInlineSnapshot(`
+      Object {
+        "products": Array [
+          Object {
+            "name": "undefined (undefined)",
```

Comment:
This part gives me the most pause.

The string with `undefined` bits is a result of the `name` resolver receiving unexpectedly missing `@requires` data, but carrying on anyway. If the `requires` fields were _nested_, this would throw.

```js
name(object, { delimeter }) {
  return `${object.title}${delimeter}(${object.year})`;
},
```

Does this mean our resolvers needs to treat all `@requires` fields as nullable? I think so! Is that unreasonable or terrible, maybe not - maybe it's the best we can do? In any case, I think we need to document this since it's a potential pitfall of any `@requires` fields that our resolvers expect to have. And with a wider lens, I think we could set the same expectations for any requests that cross service boundaries like this.

I don't think I have any real ask with this comment, but I do want to raise concerns and instigate conversation about the implications here.

Reply:
Thanks for your review @trevor-scheer . I also was wondering about this case. In this case here both fields are nullable. And I think external fields should always be. Maybe we should document this as a kind of best-practice.

Reply:
Agreed - this isn't something that's documented much in the Federation docs, but extending with non-nullable fields makes your aggregated service not fault-tolerant at all.